### DISCUSSION TITLE
JAX Jacobian with upstream gradient?

### DISCUSSION THREAD

**Discussion by karan-dalal** [jax-ml/jax#18092] [Q&A] [answered]
Hi all! Had a question regarding efficient computation of a Jacobian with an upstream gradient. I'm writing a custom backward pass using VJPs and need to compute the following.

To begin, my custom VJP receives the upstream gradient, let's label this _**dL/dz**_. I have the following function inside my custom VJP:
```python
def f(params, token):
   z = InnerNet.apply({'params': params}, token, train=False) # (Just computes linear product)
   return z
``` 

I want to compute dL/dW, where W is passed in as the params argument. I understand one way of doing this is computing the full jacobian of dz/dW and then multiply by the upstream gradient, but was wondering if there is a more efficient process without explicitly computing the complete jacobian.

At a high-level, does the autodiff cookbook have some sort of function that models the following behavior:
<img width="733" alt="Screenshot 2023-10-12 at 6 18 44 PM" src="https://github.com/google/jax/assets/44450861/e4d8b809-25d5-4bec-baaa-fd2302703515">

Thanks!

**Comment by karan-dalal** [comment]
@jakevdp any help would be great!

**Comment by patrick-kidger** [comment, marked as answer]
I think `_, vjp_fn = jax.vjp(f, params, ...)` will do what you want. You can then pass the dL/dz cotangents to `vjp_fn` and it will compute the cotangents dL/dparams.

**Reply by karan-dalal** [reply]
Many thanks Patrick!
